### DISCUSSION TITLE
How to Capture Full-Page Screenshots in SeleniumBase?

### DISCUSSION THREAD

**Discussion by dsmabulage** [seleniumbase/SeleniumBase#3362] [Q&A] [answered]
I'm currently using SeleniumBase for my automated tests, and I need to capture full-page screenshots of the pages I'm testing. However, whenever I use `driver.save_screenshot()`, the screenshots are cropped and only show the visible part of the page.

`driver.save_screenshot(f"table_element{i}.png")`

Any advice or solution would be greatly appreciated!

**Comment by mdmintz** [comment, marked as answer]
Set the window size before taking the screenshot if the current size isn't good enough:

```python
driver.set_window_size(x, y)
```

If you're running tests in headless mode, the default starting size is larger than normal.

More info: https://github.com/seleniumbase/SeleniumBase/discussions/2469#discussioncomment-8368504.

Useful examples:
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_image_saving.py
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_save_screenshots.py

The SeleniumBase formats have their own screenshot methods separate from the `driver`:
* `self.save_screenshot(name, folder=None, selector=None, by="css selector")`
* `self.save_screenshot_to_logs(name=None, selector=None, by="css selector")`
* `self.save_element_as_image_file(selector, file_name, folder=None, overlay_text="")`

Methods called directly from the `driver` use the raw `selenium` API: https://github.com/SeleniumHQ/selenium, so if you're looking for specific functionality, check those docs. (You might want to see https://github.com/SeleniumHQ/selenium/issues/14116, as that is related to your question.)

**Reply by dsmabulage** [reply]
```
import time
from seleniumbase import Driver


driver = Driver(
    uc=True,
    headless=False,
)

driver.get(
    "https://stackoverflow.com/questions/43008666/entityframeworkcore-does-not-exist-in-the-namespace-microsoft"
)

time.sleep(5)

x = driver.execute_script("return document.documentElement.scrollWidth")
y = driver.execute_script("return document.documentElement.scrollHeight")

print(x, y)

driver.set_window_size(x, y)

driver.save_screenshot(f"ss.png")

driver.quit()
```

`1249 8781`

I tried this approach but still getting cropped screenshots (only the visible part of the page). Am I missing something here?

**Reply by mdmintz** [reply]
Use this:

```python
from seleniumbase import SB

with SB(uc=True, test=True, ad_block=True, headless2=True) as sb:
    sb.set_window_size(1400, 8000)
    url = "https://stackoverflow.com/questions/43008666/entityframeworkcore-does-not-exist-in-the-namespace-microsoft"
    sb.activate_cdp_mode(url)
    sb.save_screenshot("stack.png")
```

**Reply by dsmabulage** [reply]
> Use this:
> 
> ```python
> from seleniumbase import SB
> 
> with SB(uc=True, test=True, ad_block=True, headless2=True) as sb:
>     sb.set_window_size(1400, 8000)
>     url = "https://stackoverflow.com/questions/43008666/entityframeworkcore-does-not-exist-in-the-namespace-microsoft"
>     sb.activate_cdp_mode(url)
>     sb.save_screenshot("stack.png")
> ```

Can't we execute this method without `headless=True` because there are some manual user interactions? I tried it, and it didn't work.

**Reply by mdmintz** [reply]
Set the `selector`. Eg: `sb.save_screenshot("stack.png", selector="body")`

```python
from seleniumbase import SB

with SB(uc=True, test=True, ad_block=True) as sb:
    url = "https://stackoverflow.com/questions/43008666/entityframeworkcore-does-not-exist-in-the-namespace-microsoft"
    sb.activate_cdp_mode(url)
    sb.save_screenshot("stack.png", selector="body")
```

**Comment by skald** [comment]
What is the best way to get just base64 of a screenshot with save_screenshot without saving a file? Not a binary png/jpg data. just base64
